### ISSUE TITLE
RotateShoulderUp calculation appears incorrect when forwardDistanceRatio is negative and distinctShoulderRotationLimitBackward is not 0

### ISSUE THREAD

**Issue by sclark39** [dabeschte/VRArmIK#9]
I've been working on porting your code to Unreal, and while looking through I found this code which looks wrong:

https://github.com/dabeschte/VRArmIK/blob/75d1a0a1f5316cab1c55fa6fe5826b6041ed990d/Assets/Plugins/VRArmIK/Scripts/ShoulderPoser.cs#L132-L133

forwardDistanceRatio is guaranteed to be negative here, so flipping it and then clamping with a max of 0 will ensure that the yaw will be 0 if forwardDistanceRatio is less than -0.08.

Note that although it does appear that this line is incorrect, it doesn't matter as long as distinctShoulderRotationLimitBackward is 0.
https://github.com/dabeschte/VRArmIK/blob/75d1a0a1f5316cab1c55fa6fe5826b6041ed990d/Assets/Plugins/VRArmIK/Scripts/ShoulderPoser.cs#L20

If distinctShoulderRotationLimitBackward is not 0, the way the thresholds are set up, it would look something like this... (scaled to the appropriate limits and multipliers)
![image](https://user-images.githubusercontent.com/145649/125663683-495a067c-2a0f-4ee6-a0cd-c49d0799dd4a.png)

The discontinuity would cause a shoulder pop, I imagine...

**Comment by dabeschte** [comment]
Hi! Thanks for the great analysis. That is indeed a bug which I simply did not find, probably since I did not allow the shoulders to move backwards. Does removing the "-" sign resolve the issue? I'd be happy to merge a pull request :)

**Comment by sclark39** [comment]
Yes I think it would, but I don't have a way to test it. Just saw the issue while I was reading and wanted to make sure to mention it!

If it were me I'd probably rewrite this section to unify the if statements with something along the lines of:
```
float forwardDistanceRatio = Vector3.Dot(handShoulderOffset, shoulder.transform.forward) / armLength;
float upwardDistanceRatio = Vector3.Dot(handShoulderOffset, shoulder.transform.up) / armLength;

( float distanceRatioThreshold, float shoulderRotationMultiplier ) = 
	( 0 < forwardDistanceRatio 
		? ( 0.5f, distinctShoulderRotationMultiplier )
		: ( 0.08f, distinctShoulderRotationMultiplier * 10.0f ) );

forwardDistanceRatio = 
	Mathf.Sign(forwardDistanceRatio) *
	Mathf.Max(0.0f, Mathf.Abs(forwardDistanceRatio) - distanceRatioThreshold );

targetAngle.y = 
	Mathf.Clamp(
		forwardDistanceRatio * shoulderRotationMultiplier, 
		-distinctShoulderRotationLimitBackward, distinctShoulderRotationLimitForward);
```
But that's more than just a simple bug fix, and I'm not 100% sure if that is correct C# syntax or not!

**Comment by sclark39** [comment]
I actually ended up writing this section totally differently in my port. I decided to just scale the distance ratio into [0,1] based on a threshold range for each direction, apply a EaseInOutSine to the resulting alpha to make it feel more organic, and then multiply it by the maximum rotation in whatever direction it's supposed to be going. A bit easier to tune for my purposes.

**Comment by dabeschte** [comment]
Thanks for your contribution.
I agree that EaseInOut would make it more organic. I hope I will be able to try these changes on a VR headset in the near future and to update the repository. Unfortunately, this project is not actively maintained and I am not developing in VR/Unity anymore.
